maro/simulator/scenarios/finance/business_engine.py

```diff
@@ -0,0 +1,625 @@
+        Returns:
+            dict: Metrics information.
+        """
+
+        return DocableDict(
+            METRICS_DESC,
+        )
+
+    def __del__(self):
+        """Collect resource by order."""
+
+        self._stock_pickers.clear()
+
+        for reader in self._stock_readers:
+            # Close binary reader first, so that we can clean it correctly.
+            reader.close()
+
+    def _load_configs(self):
+        """Load configurations"""
+        self._update_config_root_path(__file__)
+        with open(os.path.join(self._config_path, "config.yml")) as fp:
+            self._config = safe_load(fp)
+
+    def _load_data(self):
+        # TODO: _load_data() stock => stock
+        # Time zone we used to transfer UTC to target time zone.
+        self._time_zone = gettz(self._config["time_zone"])
+        # Trip binary reader.
+        self._stock_readers: list = []
+        self._stock_pickers: list = []
+
+        # Our stock table used to query stock by tick.
+        stock_data_root = self._config["data_path"]
+        if stock_data_root.startswith("~"):
+            stock_data_root = os.path.expanduser(stock_data_root)
+
+        stock_data_paths = []
+        for idx in range(len(self._config["stocks"])):
+            stock_data_paths.append(os.path.join(stock_data_root, f"{self._config['stocks'][idx]}.bin"))
+
+        if not os.path.exists(stock_data_root):
+            self._build_temp_data()  # TODO： implement
+
+        for idx in range(len(self._config["stocks"])):
+            self._stock_readers.append(BinaryReader(stock_data_paths[idx]))
+            print(type(self._stock_readers[-1].start_datetime), self._stock_readers[-1].end_datetime)
+            self._stock_pickers.append(
+                self._stock_readers[idx].items_tick_picker(self._start_tick, self._max_tick, time_unit="d"))
+
+    def _init_frame(self):
+        self._frame = build_frame(len(self._config["stocks"]), self.calc_max_snapshot_num())
+        self._snapshots = self._frame.snapshots
+        self._account: Account = self._frame.account[0]
+        self._account.set_init_state(init_money=self._config['account']['money'])
+        self._stocks = self._frame.stocks
+
+        for idx in range(len(self._config["stocks"])):
+            self._stocks[idx].set_init_state(self._config["stocks"][idx])
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(DECISION_EVENT, self._on_action_recieved)
+
+    def is_market_closed(self):
+        # TODO: Implement
+        return True
+
+    def _on_action_recieved(self, event: Event):
+        actions = event.payload
+        if actions is None:
+            return
+
+        for action in actions:
+            if action is None:
+                continue
+            elif isinstance(action, CancelOrder):
+                self._cancel_order(action=action, tick=event.tick)
+            else:  # Action is Order
```

Comment:
never put the important case in else!
 if OrderRequest:
elif OrderCancel:
else:
exception!

Reply:
why not use event style here?

Reply:
OrderRequestEvent + order_request_handler
OrderCancelEvent + order_cancel_handler